pkg/lightning/backend/tidb/tidb.go

```diff
@@ -364,31 +378,79 @@ func (be *tidbBackend) ImportEngine(context.Context, uuid.UUID) error {
 }
 
 func (be *tidbBackend) WriteRows(ctx context.Context, _ uuid.UUID, tableName string, columnNames []string, rows kv.Rows) error {
-	var err error
-outside:
+rowLoop:
 	for _, r := range rows.SplitIntoChunks(be.MaxChunkSize()) {
+		batch := true
+	retryLoop:
 		for i := 0; i < writeRowsMaxRetryTimes; i++ {
-			err = be.WriteRowsToDB(ctx, tableName, columnNames, r)
+			err := be.WriteRowsToDB(ctx, tableName, columnNames, r, batch)
 			switch {
 			case err == nil:
-				continue outside
+				continue rowLoop
 			case common.IsRetryableError(err):
 				// retry next loop
 			default:
-				return err
+				// retry next loop when batch is true, which means the last WriteRowsToDB failed in the batch mode,
+				// we need to redo the writing row-by-row to find where the error located and skip it correctly.
+				if batch {
+					i = 0
+					batch = false
+					continue retryLoop
+				}
+				// If batch is false and the error is not nil, it means we reach the max error count in the non-batch mode.
+				return errors.Annotatef(err, "[%s] write rows reach max error count %d", tableName, be.maxErrorCount)
 			}
 		}
-		return errors.Annotatef(err, "[%s] write rows reach max retry %d and still failed", tableName, writeRowsMaxRetryTimes)
```

Comment:
Need to keep this line if WriteRowsToDB return retryable error more than writeRowsMaxRetryTimes times